madara/crates/client/sync/src/tests/reorg.rs

```diff
@@ -0,0 +1,418 @@
+        include_str!("../../../../resources/sepolia.block_0_class_1.json"),
+    );
+    gateway_mock.mock_block_from_json(1, include_str!("../../../../resources/sepolia.block_1.json"));
+    gateway_mock.mock_class_from_json(
+        "0x1b661756bf7d16210fc611626e1af4569baa1781ffc964bd018f4585ae241c1",
+        include_str!("../../../../resources/sepolia.block_1_class_0.json"),
+    );
+    gateway_mock.mock_block_from_json(2, include_str!("../../../../resources/sepolia.block_2.json"));
+    gateway_mock.mock_class_from_json(
+        "0x4f23a756b221f8ce46b72e6a6b10ee7ee6cf3b59790e76e02433104f9a8c5d1",
+        include_str!("../../../../resources/sepolia.block_2_class_0.json"),
+    );
+    gateway_mock.mock_block_pending_not_found();
+    gateway_mock.mock_header_latest(2, felt!("0x7a906dfd1ff77a121b8048e6f750cda9e949d341c4487d4c6a449f183f0e61d"));
+
+    ReorgTestContext { backend, importer, gateway_mock }
+}
+
+#[rstest]
+#[case::reorg_from_genesis(ReorgTestArgs {
+    original_chain_length: 1,
+    orphaned_chain_length: 2,
+    new_chain_length: 1,
+    passes: 1,
+})]
+#[case::empty_orphan_chain(ReorgTestArgs {
+    original_chain_length: 2,
+    orphaned_chain_length: 0,
+    new_chain_length: 1,
+    passes: 1,
+})]
+#[case::standard_reorg(ReorgTestArgs {
+    original_chain_length: 2,
+    orphaned_chain_length: 3,
+    new_chain_length: 2,
+    passes: 1,
+})]
+#[case::multiple_passes(ReorgTestArgs {
+    original_chain_length: 1,
+    orphaned_chain_length: 3,
+    new_chain_length: 2,
+    passes: 5,
+})]
+#[case::multiple_passes_no_new_chain(ReorgTestArgs {
+    original_chain_length: 1,
+    orphaned_chain_length: 3,
+    new_chain_length: 0,
+    passes: 5,
+})]
+#[case::multiple_passes_repetitive_fork(ReorgTestArgs {
+    original_chain_length: 1,
+    orphaned_chain_length: 2,
+    new_chain_length: 2,
+    passes: 5,
+})]
+#[case::deep_reorg(ReorgTestArgs {
+    original_chain_length: 1,
+    orphaned_chain_length: 2,
+    new_chain_length: 2,
+    passes: 1,
+})]
+#[tokio::test]
+/// Comprehensive reorg test covering multiple scenarios
+///
+/// This test validates the blockchain reorganization (reorg) functionality
+/// by simulating various fork scenarios where the chain needs to revert
+/// to a common ancestor and rebuild from an alternative fork.
+///
+/// Test scenarios:
+/// 1. Reorg from genesis - Tests reverting back to block 0
+/// 2. Empty orphan chain - Tests reorg when no blocks need to be orphaned
+/// 3. Standard reorg - Tests typical reorg with one orphaned block
+/// 4. Multiple passes - Tests repeated reorgs to ensure robustness
+/// 5. Deep reorg - Tests reorging multiple blocks
+async fn test_reorg(reorg_ctx: ReorgTestContext, #[case] args: ReorgTestArgs) {
+    assert!(args.original_chain_length > 0, "Cannot create an empty chain, we always need at least genesis");
+    assert!(args.passes > 0, "Must have at least one pass");
+    assert!(
+        args.original_chain_length <= 2,
```

Comment:
there are 0, 1 and 2 blocks, which means max length would be 3?

I noticed we are doing -1 to the original_chain_length below to get the block anyway, so this makes sense!

```suggestion
        args.original_chain_length <= 3,
```

Reply:
removed

Reply:
this is not valid anymore